## Data Exploration

### Load Data

In [1]:
import pandas as pd
from config import LocalConfigurations

config = LocalConfigurations()

## Explore the four different datasets

### 1. Print Schemas

In [2]:
df_demographic = pd.read_csv(config.DEMOGRAPHICS_DATA_PATH, delimiter=";")
df_demographic.info()
df_demographic.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2891 entries, 0 to 2890
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   City                    2891 non-null   object 
 1   State                   2891 non-null   object 
 2   Median Age              2891 non-null   float64
 3   Male Population         2888 non-null   float64
 4   Female Population       2888 non-null   float64
 5   Total Population        2891 non-null   int64  
 6   Number of Veterans      2878 non-null   float64
 7   Foreign-born            2878 non-null   float64
 8   Average Household Size  2875 non-null   float64
 9   State Code              2891 non-null   object 
 10  Race                    2891 non-null   object 
 11  Count                   2891 non-null   int64  
dtypes: float64(6), int64(2), object(4)
memory usage: 271.2+ KB


City          State  Median Age  Male Population  \
0     Silver Spring       Maryland        33.8          40601.0   
1            Quincy  Massachusetts        41.0          44129.0   
2            Hoover        Alabama        38.5          38040.0   
3  Rancho Cucamonga     California        34.5          88127.0   
4            Newark     New Jersey        34.6         138040.0   

   Female Population  Total Population  Number of Veterans  Foreign-born  \
0            41862.0             82463              1562.0       30908.0   
1            49500.0             93629              4147.0       32935.0   
2            46799.0             84839              4819.0        8229.0   
3            87105.0            175232              5821.0       33878.0   
4           143873.0            281913              5829.0       86253.0   

   Average Household Size State Code                       Race  Count  
0                    2.60         MD         Hispanic or Latino  25924  
1                    2.39         MA                      White  58723  
2                    2.58         AL                      Asian   4759  
3                    3.18         CA  Black or African-American  24437  
4                    2.73         NJ                      White  76402

In [3]:
df_airport_codes = pd.read_csv(config.AIRPORT_DATA_PATH)
df_airport_codes.info()
df_airport_codes.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55075 entries, 0 to 55074
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ident         55075 non-null  object 
 1   type          55075 non-null  object 
 2   name          55075 non-null  object 
 3   elevation_ft  48069 non-null  float64
 4   continent     27356 non-null  object 
 5   iso_country   54828 non-null  object 
 6   iso_region    55075 non-null  object 
 7   municipality  49399 non-null  object 
 8   gps_code      41030 non-null  object 
 9   iata_code     9189 non-null   object 
 10  local_code    28686 non-null  object 
 11  coordinates   55075 non-null  object 
dtypes: float64(1), object(11)
memory usage: 5.0+ MB


ident           type                                name  elevation_ft  \
0   00A       heliport                   Total Rf Heliport          11.0   
1  00AA  small_airport                Aero B Ranch Airport        3435.0   
2  00AK  small_airport                        Lowell Field         450.0   
3  00AL  small_airport                        Epps Airpark         820.0   
4  00AR         closed  Newport Hospital & Clinic Heliport         237.0   

  continent iso_country iso_region  municipality gps_code iata_code  \
0       NaN          US      US-PA      Bensalem      00A       NaN   
1       NaN          US      US-KS         Leoti     00AA       NaN   
2       NaN          US      US-AK  Anchor Point     00AK       NaN   
3       NaN          US      US-AL       Harvest     00AL       NaN   
4       NaN          US      US-AR       Newport      NaN       NaN   

  local_code                            coordinates  
0        00A     -74.93360137939453, 40.07080078125  
1       00AA                 -101.473911, 38.704022  
2       00AK            -151.695999146, 59.94919968  
3       00AL  -86.77030181884766, 34.86479949951172  
4        NaN                    -91.254898, 35.6087

In [6]:
df_immigration = pd.read_csv(config.SAS_DATA_SAMPLE_PATH)
df_immigration.info()
df_immigration.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 29 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1000 non-null   int64  
 1   cicid       1000 non-null   float64
 2   i94yr       1000 non-null   float64
 3   i94mon      1000 non-null   float64
 4   i94cit      1000 non-null   float64
 5   i94res      1000 non-null   float64
 6   i94port     1000 non-null   object 
 7   arrdate     1000 non-null   float64
 8   i94mode     1000 non-null   float64
 9   i94addr     941 non-null    object 
 10  depdate     951 non-null    float64
 11  i94bir      1000 non-null   float64
 12  i94visa     1000 non-null   float64
 13  count       1000 non-null   float64
 14  dtadfile    1000 non-null   int64  
 15  visapost    382 non-null    object 
 16  occup       4 non-null      object 
 17  entdepa     1000 non-null   object 
 18  entdepd     954 non-null    object 
 19  entdepu     0 non-null      

Unnamed: 0      cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  \
0     2027561  4084316.0  2016.0     4.0   209.0   209.0     HHW  20566.0   
1     2171295  4422636.0  2016.0     4.0   582.0   582.0     MCA  20567.0   
2      589494  1195600.0  2016.0     4.0   148.0   112.0     OGG  20551.0   
3     2631158  5291768.0  2016.0     4.0   297.0   297.0     LOS  20572.0   
4     3032257   985523.0  2016.0     4.0   111.0   111.0     CHM  20550.0   

   i94mode i94addr  ...  entdepu  matflag  biryear   dtaddto  gender insnum  \
0      1.0      HI  ...      NaN        M   1955.0  07202016       F    NaN   
1      1.0      TX  ...      NaN        M   1990.0  10222016       M    NaN   
2      1.0      FL  ...      NaN        M   1940.0  07052016       M    NaN   
3      1.0      CA  ...      NaN        M   1991.0  10272016       M    NaN   
4      3.0      NY  ...      NaN        M   1997.0  07042016       F    NaN   

  airline        admnum  fltno  visatype  
0      JL  5.658267e+10  00782        WT  
1     *GA  9.436200e+10  XBLNG        B2  
2      LH  5.578047e+10  00464        WT  
3      QR  9.478970e+10  00739        B2  
4     NaN  4.232257e+10   LAND        WT  

[5 rows x 29 columns]

In [7]:
df_temperature = pd.read_csv(config.TEMPERATURE_DATA_PATH)
df_temperature.info()
df_temperature.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8599212 entries, 0 to 8599211
Data columns (total 7 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   dt                             object 
 1   AverageTemperature             float64
 2   AverageTemperatureUncertainty  float64
 3   City                           object 
 4   Country                        object 
 5   Latitude                       object 
 6   Longitude                      object 
dtypes: float64(2), object(5)
memory usage: 459.2+ MB


dt  AverageTemperature  AverageTemperatureUncertainty   City  \
0  1743-11-01               6.068                          1.737  Århus   
1  1743-12-01                 NaN                            NaN  Århus   
2  1744-01-01                 NaN                            NaN  Århus   
3  1744-02-01                 NaN                            NaN  Århus   
4  1744-03-01                 NaN                            NaN  Århus   

   Country Latitude Longitude  
0  Denmark   57.05N    10.33E  
1  Denmark   57.05N    10.33E  
2  Denmark   57.05N    10.33E  
3  Denmark   57.05N    10.33E  
4  Denmark   57.05N    10.33E

### 2. Extract Unique values


In [8]:
def extract_info(df, n):
    countLabel = "Frequency"
    print(f"Number of values for this table is {df.shape[0]}")
    for col in df:
        df[countLabel] = df.groupby(col)[col].transform('count')
        df.sort_values(countLabel, inplace=True, ascending=False)
        counter = df[[col, countLabel]]
        counter = counter.drop_duplicates()
        print(f"Column: {col}")
        print(f"Total number of unique values are {len(df[col].unique())}")
        print(f"{n} most frequent values")
        print(f"{counter.head(n)}")
        print(f"Number of NaN values: {df[col].isna().sum()}, {df[col].isna().sum()/df.shape[0] * 100}%)")
        print("______________________")
    df.drop(columns=[countLabel], inplace=True)

#### Demographic

Check how many rows have NaN and the most frequent values for each column

In [9]:
extract_info(df_demographic, 10)

Number of values for this table is 2891
Column: City
Total number of unique values are 567
10 most frequent values
             City  Frequency
2167     Columbia         15
725   Springfield         15
549   Bloomington         15
2714      Jackson         10
1621       Aurora         10
668    Wilmington         10
227      Glendale         10
1016     Richmond         10
890       Norwalk         10
1261     Pasadena         10
Number of NaN values: 0, 0.0%)
______________________
Column: State
Total number of unique values are 49
10 most frequent values
               State  Frequency
2426      California        676
1734           Texas        273
2270         Florida        222
22          Illinois         91
274       Washington         85
630          Arizona         80
26          Colorado         80
2781        Michigan         79
1162        Virginia         70
1762  North Carolina         70
Number of NaN values: 0, 0.0%)
______________________
Column: Median Age
Total number

When exploring the data above the most interesting points are the following

1. Very low number of rows have NaN values in any of the columns
2. It is not very clear what the column Count represents

First check if Count matches the total populations when the count has been aggregated over all races. 

In [10]:
df_demographic.groupby(["City", "State", "Median Age", "Female Population", "Male Population", "Total Population", "Number of Veterans", "Foreign-born", "Average Household Size"], dropna=False).sum("Count").head(15)

Count
City        State        Median Age Female Population Male Population Total Population Number of Veterans Foreign-born Average Household Size        
Abilene     Texas        31.3       60664.0           65212.0         125876           9367.0             8129.0       2.64                    147900
Akron       Ohio         38.1       100667.0          96886.0         197553           12878.0            10024.0      2.24                    210305
Alafaya     Florida      33.5       45760.0           39504.0         85264            4176.0             15842.0      2.94                    115476
Alameda     California   41.4       40867.0           37747.0         78614            4504.0             18841.0      2.52                     89174
Albany      Georgia      33.3       39414.0           31695.0         71109            5409.0             861.0        2.38                     73478
            New York     32.8       50825.0           47627.0         98452            3643.0             11948.0      2.08                    108740
Albuquerque New Mexico   36.0       285808.0          273323.0        559131           37443.0            58200.0      2.49                    767858
Alexandria  Virginia     36.6       78522.0           74989.0         153511           10635.0            44030.0      2.20                    183404
Alhambra    California   41.0       43388.0           42184.0         85572            1673.0             44441.0      2.89                     98856
Allen       Pennsylvania 33.5       59581.0           60626.0         120207           5691.0             19652.0      2.67                    159413
            Texas        37.2       46814.0           51324.0         98138            3505.0             19649.0      3.04                    109612
Amarillo    Texas        33.8       100260.0          99391.0         199651           11008.0            21124.0      2.64                    266479
Ames        Iowa         23.0       31238.0           33814.0         65052            2265.0             8606.0       2.16                     68263
Anaheim     California   33.6       171135.0          179603.0        350738           9796.0             137133.0     3.45                    526947
Anchorage   Alaska       32.2       145750.0          152945.0        298695           27492.0            33258.0      2.77                    336228

That is not the case, so Count is still very confusing. Let's all check if there can exist a city, state pair with different value of age, populations etc.

In [11]:
df_demographic.groupby(["City", "State"], dropna=False).std().max(axis=0, numeric_only=True)

Median Age                0.000000e+00
Male Population           0.000000e+00
Female Population         0.000000e+00
Total Population          0.000000e+00
Number of Veterans        0.000000e+00
Foreign-born              0.000000e+00
Average Household Size    0.000000e+00
Count                     1.393622e+06
dtype: float64

As expected that is not the case, so let's now look at the NaN rows.

In [12]:
nan_df_demo = df_demographic[df_demographic.isna().any(axis=1)]
print(f"Number of NaN rows is {nan_df_demo.shape[0]}")
nan_df_demo.head(20)

Number of NaN rows is 16


City        State  Median Age  Male Population  \
2441      San Juan  Puerto Rico        41.4         155408.0   
1748      Mayagüez  Puerto Rico        38.1          30799.0   
449   The Villages      Florida        70.5              NaN   
2597        Caguas  Puerto Rico        40.4          34743.0   
258       Carolina  Puerto Rico        42.0          64758.0   
2654      Guaynabo  Puerto Rico        42.2          33066.0   
1747      San Juan  Puerto Rico        41.4         155408.0   
2746      Mayagüez  Puerto Rico        38.1          30799.0   
2589      Guaynabo  Puerto Rico        42.2          33066.0   
2004       Bayamón  Puerto Rico        39.4          80128.0   
155         Caguas  Puerto Rico        40.4          34743.0   
1995         Ponce  Puerto Rico        40.5          56968.0   
111       San Juan  Puerto Rico        41.4         155408.0   
637       Carolina  Puerto Rico        42.0          64758.0   
1437  The Villages      Florida        70.5              NaN   
333   The Villages      Florida        70.5              NaN   

      Female Population  Total Population  Number of Veterans  Foreign-born  \
2441           186829.0            342237                 NaN           NaN   
1748            35782.0             66581                 NaN           NaN   
449                 NaN             72590             15231.0        4034.0   
2597            42265.0             77008                 NaN           NaN   
258             77308.0            142066                 NaN           NaN   
2654            37426.0             70492                 NaN           NaN   
1747           186829.0            342237                 NaN           NaN   
2746            35782.0             66581                 NaN           NaN   
2589            37426.0             70492                 NaN           NaN   
2004            90131.0            170259                 NaN           NaN   
155             42265.0             77008                 NaN           NaN   
1995            64615.0            121583                 NaN           NaN   
111            186829.0            342237                 NaN           NaN   
637             77308.0            142066                 NaN           NaN   
1437                NaN             72590             15231.0        4034.0   
333                 NaN             72590             15231.0        4034.0   

      Average Household Size State Code                               Race  \
2441                     NaN         PR                              Asian   
1748                     NaN         PR                              Asian   
449                      NaN         FL          Black or African-American   
2597                     NaN         PR  American Indian and Alaska Native   
258                      NaN         PR  American Indian and Alaska Native   
2654                     NaN         PR  American Indian and Alaska Native   
1747                     NaN         PR  American Indian and Alaska Native   
2746                     NaN         PR                 Hispanic or Latino   
2589                     NaN         PR                 Hispanic or Latino   
2004                     NaN         PR                 Hispanic or Latino   
155                      NaN         PR                 Hispanic or Latino   
1995                     NaN         PR                 Hispanic or Latino   
111                      NaN         PR                 Hispanic or Latino   
637                      NaN         PR                 Hispanic or Latino   
1437                     NaN         FL                              White   
333                      NaN         FL                 Hispanic or Latino   

       Count  
2441    2452  
1748     235  
449      331  
2597     624  
258    12143  
2654     589  
1747    4031  
2746   65521  
2589   69936  
2004  169155  
155    76349  
1995  120705  
111   335559  
637   139967  
1437   72211  
333     1066

As can be seen most of the missing values are for cities within Puerto Rico and then The Villages is then a census-designated place of only 72590 people, so these row do not have a valueable information when it comes to places in the U.S. for immigrations.

#### airport info

Check how many rows have NaN and the most frequent values for each column

In [13]:
extract_info(df_airport_codes, 10)

Number of values for this table is 55075
Column: ident
Total number of unique values are 55075
10 most frequent values
      ident  Frequency
0       00A          1
36708  NVSW          1
36710  NVSZ          1
36711  NVVA          1
36712  NVVB          1
36713  NVVD          1
36714  NVVF          1
36715  NVVI          1
36716  NVVJ          1
36717  NVVK          1
Number of NaN values: 0, 0.0%)
______________________
Column: type
Total number of unique values are 7
10 most frequent values
                 type  Frequency
55074   small_airport      33965
23670        heliport      11287
31420  medium_airport       4550
40650          closed       3606
383     seaplane_base       1016
30038   large_airport        627
36825     balloonport         24
Number of NaN values: 0, 0.0%)
______________________
Column: name
Total number of unique values are 52144
10 most frequent values
                                 name  Frequency
21053     Centre Hospitalier Heliport         85
28945   

As can be seen there are few columns that have quite high number of NaN values, let's look at few ones.

In [14]:
df_us = df_airport_codes.loc[df_airport_codes['iso_country'] == "US"]
print(f"Values total {df_us.shape[0]} with {df_us[df_us['continent'].isna()].shape[0]} continent NaN rows ")
df_us.head(10)

Values total 22757 with 22756 continent NaN rows 


ident           type                                       name  \
49898  US-0883  large_airport                                        JFK   
49947  US-0932  small_airport                                        CLE   
49999  US-0984  small_airport                                        0c2   
49820  US-0805  small_airport                                Twin Cities   
50032  US-1016  large_airport                                        JFK   
26999     KEDC  small_airport                   Austin Executive Airport   
49828  US-0813  small_airport  (Duplicate)Grandpa's Farm Mendota Airport   
38982     PN05       heliport                          Fort Lee Heliport   
25518     K3L2  small_airport                          Sky Ranch Airport   
36637     NV21       heliport                         Sky Ranch Heliport   

       elevation_ft continent iso_country iso_region   municipality gps_code  \
49898           NaN       NaN          US      US-NY       New York      NaN   
49947           NaN       NaN          US     US-U-A      Cleveland      NaN   
49999           NaN       NaN          US      US-IL            NaN      NaN   
49820           NaN       NaN          US     US-U-A  Tabor City,NC     K5J9   
50032           NaN       NaN          US      US-NY  New York City      NaN   
26999         620.0       NaN          US      US-TX         Austin     KEDC   
49828         727.0       NaN          US      US-IL        Mendota     IL22   
38982        1995.0       NaN          US      US-PA      Barnsboro     PN05   
25518        2599.0       NaN          US      US-NV   Sandy Valley     K3L2   
36637        2600.0       NaN          US      US-NV   Sandy Valley     NV21   

      iata_code local_code                            coordinates  
49898       NaN        NaN                                   0, 0  
49947       NaN        NaN                                   0, 0  
49999       NaN        NaN                                   0, 0  
49820       NaN        NaN                                   0, 0  
50032       NaN        NaN                                   0, 0  
26999       NaN        EDC                -97.5663935, 30.3974931  
49828       NaN       IL22                     0.000001, 0.000001  
38982       NaN       PN05  -78.82720184326172, 40.65140151977539  
25518       NaN        3L2  -115.62699890136719, 35.7952995300293  
36637       NaN       NV21  -115.62699890136719, 35.7952995300293

In [15]:
# let's check what airport in the U.S. does have a continent value
print(df_us[df_us['continent'].notnull()])
df_airport_codes.loc[df_airport_codes['continent'] == "AS"].head()

      ident      type           name  elevation_ft continent iso_country  \
37387  OI05  heliport  Dana Heliport         625.0        AS          US   

      iso_region municipality gps_code iata_code local_code  \
37387      US-OH       Toledo     OI05       NaN       OI05   

                             coordinates  
37387  -83.6465988159, 41.65589904789999  


ident           type                         name  elevation_ft  \
38734  PH-0165  small_airport  clark international airport           NaN   
38736  PH-0167  large_airport                         NAIA           NaN   
31531      LHR         closed               marked as spam           NaN   
18308     EFDD  large_airport                   LOAN OFFER           NaN   
38735  PH-0166  large_airport                          MNL           NaN   

      continent iso_country iso_region           municipality gps_code  \
38734        AS          PH     PH-PAM  san fernando pampanga      NaN   
38736        AS          PH     PH-U-A             Pasay City      NaN   
31531        AS          AF     AF-U-A                    NaN      NaN   
18308        AS          AE      AE-DU                  dubai     EFDD   
38735        AS          PH     PH-TAR            Tarlac City      NaN   

      iata_code local_code coordinates  
38734       NaN        NaN        0, 0  
38736       NaN        NaN        0, 0  
31531       NaN       0000        0, 0  
18308       LHG       NY33        0, 0  
38735       NaN        NaN        0, 0

AS stands for Asia so the continent value for the only airport in the U.S. that had a value was wrong.

Also earlier when we extracted the most common airport continent we saw that North America was missing as a continent

Cont.  Count  Name
EU     7840   Europe
SA     7709   South America
AS     5350   Asia
AF     3362   Africa
OC     3067   Oceania
AN       28   Antarctica
NaN   27719   Unknown

In [16]:
df_airport_codes.loc[df_airport_codes['name'] == 'JFK'].head()

ident           type name  elevation_ft continent iso_country  \
49898  US-0883  large_airport  JFK           NaN       NaN          US   
50032  US-1016  large_airport  JFK           NaN       NaN          US   

      iso_region   municipality gps_code iata_code local_code coordinates  
49898      US-NY       New York      NaN       NaN        NaN        0, 0  
50032      US-NY  New York City      NaN       NaN        NaN        0, 0

Here we can see that there are some major airports in the U.S. missing local code. In addition all the US airports are missing continent value.

#### temperature
Check how many rows have NaN and the most frequent values for each column

In [17]:
extract_info(df_temperature, 10)

Number of values for this table is 8599212
Column: dt
Total number of unique values are 3239
10 most frequent values
                 dt  Frequency
8599211  2013-09-01       3510
4999244  2007-07-01       3510
4999254  2008-05-01       3510
4999253  2008-04-01       3510
4999252  2008-03-01       3510
4999251  2008-02-01       3510
4999250  2008-01-01       3510
4999249  2007-12-01       3510
4999248  2007-11-01       3510
4999247  2007-10-01       3510
Number of NaN values: 0, 0.0%)
______________________
Column: AverageTemperature
Total number of unique values are 103482
10 most frequent values
         AverageTemperature  Frequency
4933570              26.544      828.0
5146348              26.007      797.0
6078772              26.045      778.0
1463719              26.612      765.0
7261822              26.091      757.0
4551843              26.063      754.0
6608163              25.773      749.0
7932367              26.470      747.0
1595242              25.843      738.0
166555

Only 4% of temperature data is missing values. Let's check how many values from 2016 are missing temperature values.

In [18]:
df_temperature.head()
df_temperature["Year"] = pd.to_datetime(df_temperature["dt"], format='%Y-%m-%d').dt.year
df_temperature_2016 = df_temperature.loc[df_temperature['Year'] == "2016"]
df_temperature_2016us = df_temperature_2016.loc[df_temperature_2016['Country'] == "United States"]
print(f"Number of records from 2016 are {df_temperature_2016.shape[0]}")
print(f"Number of records from the U.S. and from 2016 are {df_temperature_2016us.shape[0]}")

df_temperature_us = df_temperature.loc[df_temperature['Country'] == "United States"]
print(f"Number of records from the U.S. {df_temperature_us.shape[0]}")
print(f"Newest year {df_temperature['Year'].max()}")


Number of records from 2016 are 0
Number of records from the U.S. and from 2016 are 0
Number of records from the U.S. 687289
Newest year 2013


There is no data from the year 2016.

#### Immigration data
Check how many rows have NaN and the most frequent values for each column

In [19]:
extract_info(df_immigration, 10)

Number of values for this table is 1000
Column: Unnamed: 0
Total number of unique values are 1000
10 most frequent values
     Unnamed: 0  Frequency
0       2027561          1
671      439311          1
658     2597228          1
659      822854          1
660     2769197          1
661     1797128          1
662     2360091          1
663     2466616          1
664     1434117          1
665       27374          1
Number of NaN values: 0, 0.0%)
______________________
Column: cicid
Total number of unique values are 1000
10 most frequent values
        cicid  Frequency
0   4084316.0          1
13  4916639.0          1
30  5692439.0          1
29  2826530.0          1
2   1195600.0          1
3   5291768.0          1
4    985523.0          1
5   1481650.0          1
6   2197173.0          1
7    232708.0          1
Number of NaN values: 0, 0.0%)
______________________
Column: i94yr
Total number of unique values are 1
10 most frequent values
    i94yr  Frequency
0  2016.0       1000
Numbe

Column: admnum
Total number of unique values are 1000
10 most frequent values
           admnum  Frequency
535  9.405619e+10          1
636  9.383637e+10          1
989  5.583149e+10          1
856  5.610961e+10          1
268  9.412380e+10          1
270  6.754231e+08          1
110  5.570177e+10          1
296  9.266934e+10          1
456  5.633815e+10          1
24   5.621703e+10          1
Number of NaN values: 0, 0.0%)
______________________
Column: fltno
Total number of unique values are 503
10 most frequent values
     fltno  Frequency
305   LAND       19.0
233  00001       12.0
8    00008       12.0
586  00283        7.0
150  00017        7.0
633  00102        7.0
933  00003        7.0
17   08704        6.0
375  00073        6.0
590  00452        6.0
Number of NaN values: 8, 0.8%)
______________________
Column: visatype
Total number of unique values are 10
10 most frequent values
    visatype  Frequency
141       WT        443
587       B2        356
795       WB         91
151

Occupations and entdepu columns are missing a lot of values, therefore, those two columns will not be used.

Let's check what addresses have NaN

In [20]:
df_immigration[df_immigration['i94addr'].isna()][["i94port"]].head(10)

i94port
792     HHW
457     NEW
46      SDP
766     PSP
771     NYC
838     HHW
331     HHW
187     HHW
945     HHW
71      NEW

In [21]:
df_immigration[["i94bir", "biryear"]].head(10)

i94bir  biryear
141    55.0   1961.0
634    21.0   1995.0
317    22.0   1994.0
802    28.0   1988.0
355    66.0   1950.0
737    38.0   1978.0
108    48.0   1968.0
812    19.0   1997.0
498    48.0   1968.0
119    25.0   1991.0

In [22]:

df_airport_dedup = df_airport_codes.drop_duplicates(subset=['local_code'])
df_immairp = pd.merge(df_immigration,
                 df_airport_dedup,
                 left_on='i94port',
                 right_on='local_code',
                 how='inner')
print(f"Matches are {df_immairp.shape[0]}/{df_immigration.shape[0]}={df_immairp.shape[0]/df_immigration.shape[0]}")
df_immairp.head()

Matches are 652/1000=0.652


Unnamed: 0      cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  \
0     2592682  5245089.0  2016.0     4.0   148.0   112.0     CLT  20572.0   
1     1439326  2913751.0  2016.0     4.0   148.0   112.0     CLT  20560.0   
2     2700575  5454506.0  2016.0     4.0   135.0   135.0     CLT  20573.0   
3     1474254  3001596.0  2016.0     4.0   577.0   577.0     CLT  20560.0   
4     2317991  4709156.0  2016.0     4.0   148.0   112.0     CLT  20569.0   

   i94mode i94addr  ...                     name  elevation_ft  continent  \
0      1.0      KY  ...  Colonia Catriel Airport        1026.0         SA   
1      1.0      FL  ...  Colonia Catriel Airport        1026.0         SA   
2      1.0      CA  ...  Colonia Catriel Airport        1026.0         SA   
3      1.0      CA  ...  Colonia Catriel Airport        1026.0         SA   
4      1.0      SC  ...  Colonia Catriel Airport        1026.0         SA   

   iso_country  iso_region     municipality gps_code iata_code local_code  \
0           AR        AR-R  Colonia Catriel     SA30       CCT        CLT   
1           AR        AR-R  Colonia Catriel     SA30       CCT        CLT   
2           AR        AR-R  Colonia Catriel     SA30       CCT        CLT   
3           AR        AR-R  Colonia Catriel     SA30       CCT        CLT   
4           AR        AR-R  Colonia Catriel     SA30       CCT        CLT   

                      coordinates  
0  -67.8349990845, -37.9101982117  
1  -67.8349990845, -37.9101982117  
2  -67.8349990845, -37.9101982117  
3  -67.8349990845, -37.9101982117  
4  -67.8349990845, -37.9101982117  

[5 rows x 41 columns]

Read in SAS Description

In [23]:
from utils import read_sas_description
label_desc_map = read_sas_description(config.SAS_DESC_DATA_PATH)
print(label_desc_map)

{'I94YR': {}, 'I94MON': {}, 'I94CIT': {582: 'MEXICO Air Sea, and Not Reported (I-94, no land arrivals)', 236: 'AFGHANISTAN', 101: 'ALBANIA', 316: 'ALGERIA', 102: 'ANDORRA', 324: 'ANGOLA', 529: 'ANGUILLA', 518: 'ANTIGUA-BARBUDA', 687: 'ARGENTINA', 151: 'ARMENIA', 532: 'ARUBA', 438: 'AUSTRALIA', 103: 'AUSTRIA', 152: 'AZERBAIJAN', 512: 'BAHAMAS', 298: 'BAHRAIN', 274: 'BANGLADESH', 513: 'BARBADOS', 104: 'BELGIUM', 581: 'BELIZE', 386: 'BENIN', 509: 'BERMUDA', 153: 'BELARUS', 242: 'BHUTAN', 688: 'BOLIVIA', 717: 'BONAIRE, ST EUSTATIUS, SABA', 164: 'BOSNIA-HERZEGOVINA', 336: 'BOTSWANA', 689: 'BRAZIL', 525: 'BRITISH VIRGIN ISLANDS', 217: 'BRUNEI', 105: 'BULGARIA', 393: 'BURKINA FASO', 243: 'BURMA', 375: 'BURUNDI', 310: 'CAMEROON', 326: 'CAPE VERDE', 526: 'CAYMAN ISLANDS', 383: 'CENTRAL AFRICAN REPUBLIC', 384: 'CHAD', 690: 'CHILE', 245: 'CHINA, PRC', 721: 'CURACAO', 270: 'CHRISTMAS ISLAND', 271: 'COCOS ISLANDS', 691: 'COLOMBIA', 317: 'COMOROS', 385: 'CONGO', 467: 'COOK ISLANDS', 575: 'COSTA RICA

In [24]:
for key, val in label_desc_map.items():
    print(key)
    print("--------------")
    for k, v in val.items():
        print(k, v)

I94YR
--------------
I94MON
--------------
I94CIT
--------------
582 MEXICO Air Sea, and Not Reported (I-94, no land arrivals)
236 AFGHANISTAN
101 ALBANIA
316 ALGERIA
102 ANDORRA
324 ANGOLA
529 ANGUILLA
518 ANTIGUA-BARBUDA
687 ARGENTINA
151 ARMENIA
532 ARUBA
438 AUSTRALIA
103 AUSTRIA
152 AZERBAIJAN
512 BAHAMAS
298 BAHRAIN
274 BANGLADESH
513 BARBADOS
104 BELGIUM
581 BELIZE
386 BENIN
509 BERMUDA
153 BELARUS
242 BHUTAN
688 BOLIVIA
717 BONAIRE, ST EUSTATIUS, SABA
164 BOSNIA-HERZEGOVINA
336 BOTSWANA
689 BRAZIL
525 BRITISH VIRGIN ISLANDS
217 BRUNEI
105 BULGARIA
393 BURKINA FASO
243 BURMA
375 BURUNDI
310 CAMEROON
326 CAPE VERDE
526 CAYMAN ISLANDS
383 CENTRAL AFRICAN REPUBLIC
384 CHAD
690 CHILE
245 CHINA, PRC
721 CURACAO
270 CHRISTMAS ISLAND
271 COCOS ISLANDS
691 COLOMBIA
317 COMOROS
385 CONGO
467 COOK ISLANDS
575 COSTA RICA
165 CROATIA
584 CUBA
218 CYPRUS
140 CZECH REPUBLIC
723 FAROE ISLANDS (PART OF DENMARK)
108 DENMARK
322 DJIBOUTI
519 DOMINICA
585 DOMINICAN REPUBLIC
240 EAST TIMOR
692 ECUA

PIA GREATER PEORIA, IL
RFD GREATER ROCKFORD, IL
UGN MEMORIAL - WAUKEGAN, IL
GAR GARY, IN
HMM HAMMOND, IN
INP INDIANAPOLIS, IN
MRL MERRILLVILLE, IN
SBN SOUTH BEND, IN
ICT MID-CONTINENT - WITCHITA, KS
LEX BLUE GRASS - LEXINGTON, KY
LOU LOUISVILLE, KY
BTN BATON ROUGE, LA
LKC LAKE CHARLES, LA
LAK LAKE CHARLES, LA (BPS)
MLU MONROE, LA
MGC MORGAN CITY, LA
NOL NEW ORLEANS, LA
BOS BOSTON, MA
GLO GLOUCESTER, MA
BED HANSCOM FIELD - BEDFORD, MA
LYN LYNDEN, WA
ADW ANDREWS AFB, MD
BAL BALTIMORE, MD
MKG MUSKEGON, MD
PAX PATUXENT RIVER, MD
BGM BANGOR, ME
BOO BOOTHBAY HARBOR, ME
BWM BRIDGEWATER, ME
BCK BUCKPORT, ME
CLS CALAIS, ME
CRB CARIBOU, ME
COB COBURN GORE, ME
EST EASTCOURT, ME
EPT EASTPORT MUNICIPAL, ME
EPM EASTPORT, ME
FOR FOREST CITY, ME
FTF FORT FAIRFIELD, ME
FTK FORT KENT, ME
HML HAMIIN, ME
HTM HOULTON, ME
JKM JACKMAN, ME
KAL KALISPEL, MT
LIM LIMESTONE, ME
LUB LUBEC, ME
MAD MADAWASKA, ME
POM PORTLAND, ME
RGM RANGELEY, ME (BPS)
SBR SOUTH BREWER, ME
SRL ST AURELIE, ME
SPA ST PAMPILE, ME
VNB VA

CPX No PORT Code (CPX)
MTH No PORT Code (MTH)
PFN No PORT Code (PFN)
SCH No PORT Code (SCH)
ASI No PORT Code (ASI)
BKF No PORT Code (BKF)
DAY No PORT Code (DAY)
Y62 No PORT Code (Y62)
AG No PORT Code (AG)
BCM No PORT Code (BCM)
DEC No PORT Code (DEC)
PLB No PORT Code (PLB)
CXO No PORT Code (CXO)
JBQ No PORT Code (JBQ)
JIG No PORT Code (JIG)
OGS No PORT Code (OGS)
TIW No PORT Code (TIW)
OTS No PORT Code (OTS)
AMT No PORT Code (AMT)
EGE No PORT Code (EGE)
GPI No PORT Code (GPI)
NGL No PORT Code (NGL)
OLM No PORT Code (OLM)
CLX No PORT Code (CLX)
FSC No PORT Code (FSC)
NK No PORT Code (NK)
ADU No PORT Code (ADU)
AKT No PORT Code (AKT)
LIT No PORT Code (LIT)
A2A No PORT Code (A2A)
OSN No PORT Code (OSN)
I94MODE
--------------
1 Air
2 Sea
3 Land
9 Not reported ;
I94ADDR
--------------
I94BIR
--------------
I94VISA
--------------
1 Business
2 Pleasure
3 Student
COUNT
--------------
DTADFILE
--------------
VISAPOST
--------------
OCCUP
--------------
ENTDEPA
--------------
ENTDEPD
-----------